web/source/osk/inputEventCoordinate.ts

```diff
@@ -0,0 +1,51 @@
+namespace com.keyman.osk {
+  export class InputEventCoordinate {
+    public readonly x: number;
+    public readonly y: number;
+
+    private readonly source: MouseEvent | TouchEvent;
+
+    public constructor(x: number, y: number, source?: MouseEvent | TouchEvent) {
+      this.x = x;
+      this.y = y;
+
+      if(source) {
+        this.source = source;
+      }
+    }
+
+    // Converts a MouseEvent or TouchEvent into the base coordinates needed 
+    // by the mouse-dragging operations.
+    public static fromEvent(e: MouseEvent | TouchEvent) {
+      let coordSource: MouseEvent | Touch;
+      if(e instanceof TouchEvent) {
+        coordSource = e.changedTouches[0];
+      } else {
+        coordSource = e;
+      }
+
+      if (coordSource.pageX) {
+        return new InputEventCoordinate(coordSource.pageX, coordSource.pageY, e);
+      } else if (coordSource.clientX) {
+        const x = coordSource.clientX + document.body.scrollLeft;
+        const y = coordSource.clientY + document.body.scrollTop;
+
+        return new InputEventCoordinate(x, y, e);
```

Comment:
What are the scenarios in which `pageX` vs `clientX` vs none are available? Can you add a code comment?

Reply:
That is a good question; this was written to maintain existing behavior in this regard.

Based on data from caniuse.com, _for touch events_, the two seem to have equal levels of support.  This does not hold _for mouse events_ though, and this type (and function) will be used for both.

![image](https://user-images.githubusercontent.com/25213402/129651720-4079051e-007c-4bfe-8f19-f850caba1344.png)

vs

![image](https://user-images.githubusercontent.com/25213402/129651760-8ea3689c-c6de-4079-87e8-678f2bdc268a.png)

I can add a comment about _that_ aspect, at least, but I can't speak well to the intent of the original code from which this section was adapted.

Reply:
Those version numbers look reasonably safe - Chrome 44 being the only possible outlier for `pageX`, but given it's only for `MouseEvent`, I think it's still okay (we really only see old Chrome on touch devices).

Reply:
> What are the scenarios in which `pageX` vs `clientX` vs none are available? Can you add a code comment?

Reminder to self:  should probably add this in as well before the merge.